### PR TITLE
[FLINK-5903][YARN]respect taskmanager.numberOfTaskSlots and yarn.containers.vcores in YARN mode

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
Make sure taskmanager.numberOfTaskSlots and yarn.containers.vcores works in YARN mode. The priorities is: -s/-ys > yarn.containers.vcores > taskmanager.numberOfTaskSlots.

Verify the modification:

|                               | yarn session | yarn cluster |
| ----------------------------- | ------------ | ------------ |
| -s                            | ok           | /            |
| -ys                           | /            | ok           |
| yarn.containers.vcores        | ok           | ok           |
| taskmanager.numberOfTaskSlots | ok           | ok           |

And I've tried cross verification with any two of these three configs, the priorities worked right.

**Updates:**

After checking with @tillrohrmann, we came out with the conclusion that `-s/-ys` and `taskmanager.numberOfTaskSlots` is used to set #slots of TM while `yarn.containers.vcores` set #vcores of TM container and it will use values of `-s/ys` or `taskmanager.numberOfTaskSlots` if user doesn't set it.

After changing codes, the test results is like:

| Configurations                           | #vcores of container(TM) | #slots of TM |
| ---------------------------------------- | ------------------------ | ------------ |
| -s/-ys 5,       yarn.containers.vcores: 4,  taskmanager.numberOfTaskSlots: 3 | 4                        | 5            |
| yarn.containers.vcores: 4,  taskmanager.numberOfTaskSlots: 3 | 4                        | 3            |
| yarn.containers.vcores: 4                | 4                        | 1            |
| -s/-ys 5,  taskmanager.numberOfTaskSlots: 3 | 5                        | 5            |
| taskmanager.numberOfTaskSlots: 3         | 3                        | 3            |
| Nothing to specify, all use default      | 1                        | 1            |

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 1]
```diff
@@ -317,6 +319,10 @@ public AbstractYarnClusterDescriptor createDescriptor(String defaultApplicationN
 		if (cmd.hasOption(SLOTS.getOpt())) {
 			int slots = Integer.valueOf(cmd.getOptionValue(SLOTS.getOpt()));
 			yarnClusterDescriptor.setTaskManagerSlots(slots);
+		} else if (config.containsKey(ConfigConstants.YARN_VCORES)) {
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: I think the preference order should be `TASK_MANAGER_NUM_TASK_SLOTS` > `YARN_VCORES` for the slots.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 2]
```diff
@@ -537,7 +543,6 @@ public YarnClusterClient createCluster(
 		Preconditions.checkNotNull(userJarFiles, "User jar files should not be null.");
 
 		AbstractYarnClusterDescriptor yarnClusterDescriptor = createDescriptor(applicationName, cmdLine);
-		yarnClusterDescriptor.setFlinkConfiguration(config);
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: You cannot simply ignore the `configuration` which has been given to `createCluster`. What if it contains configuration values which were added programmatically?

Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: You should give this configuration to the `createDescriptor`  method where it is used to instantiate the `YarnClusterDescriptor`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Actually, I'm not so sure whether whether we should use the `YARN_VCORES` value as a fallback value for the number of task manager slots. We had this in the past but we changed it so that the default behaviour is if the number of slots is not specified, then it is `1`. Changing this here for Yarn would make the behaviour inconsistent wrt Mesos, for example. From a user's perspective the old behaviour is imo more predictable. However, we should definitely respect the configuration file settings for the task manager slots.

**Comment by tillrohrmann** [comment]
Another remark concerning the commit message. In the Flink community it's common to format the commit message the following way: 
```
[FLINK-XXXX] [component tag] Commit message title

Commit message body
```

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (shijinkui) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: @tillrohrmann I'm agree with you. We need make sense that YARN_VCORES is available  in yarn mode, and yarn/mesos/standalone should have different configuration. And `YARN_VCORES` is the sum of `Resource#getVirtualCores` response from yarn client.

Comment (WangTaoTheTonic) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: @tillrohrmann You mean YARN_VCORES is deprecated now? After checking code I found there're two places where we still use it: [here](https://github.com/apache/flink/blob/master/flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java#L317) and [here](https://github.com/apache/flink/blob/master/flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java#L339), especially the latter one is used for container request.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 2 above)
Comment (WangTaoTheTonic) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: @tillrohrmann Under what condition this configuration will contains values added programmatically? I've checked the codes and only found this config is initiallized from config file.

**Comment by WangTaoTheTonic** [comment]
I'm sorry about the commit message :(
next time I'll format it, as it's better not to squash.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (WangTaoTheTonic) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: And in document, there still has introduction of YARN_VCORES.

```
yarn.containers.vcores The number of virtual cores (vcores) per YARN container. By default, the 
number of vcores is set to the number of slots per TaskManager, if set, or to 1, otherwise.
```

Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: No YARN_VCORES is not deprecated. But I'm not sure whether we should use this value as a default value for the number of task manager slots.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: This might be the case right now, but it could also change in the future. The signature of the `createCluster` method contains the configuration parameter and, thus, every user will think he can pass a configuration  for the creation. We should keep that way.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (WangTaoTheTonic) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: the document says YARN_VCORE > slotsOfTaskManager > default value(1) shen they are set in file.
the parameters in command line will be used before those in config file, in some way it is a common sense :)

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 2 above)
Comment (WangTaoTheTonic) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: all right

**Comment by WangTaoTheTonic** [comment]
I move the initiallization of this config to constructor of cluster descripter and restore the deleted configuration setting.
Please check if we are good with the usage of `YARN_VCORES`.

**Comment by WangTaoTheTonic** [comment]
ping @tillrohrmann

**Comment by WangTaoTheTonic** [comment]
@tillrohrmann @StephanEwen 

The code was changed and I've verified the functions, could you please review this and merge it if it's good to go?

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 3]
```diff
@@ -164,6 +164,12 @@ public AbstractYarnClusterDescriptor() {
 				throw new RuntimeException("Unable to locate configuration file in " + confFile);
 			}
 			flinkConfigurationPath = new Path(confFile.getAbsolutePath());
+
+			if (flinkConfiguration.containsKey(ConfigConstants.YARN_VCORES)) {
+				slots = flinkConfiguration.getInteger(ConfigConstants.YARN_VCORES, -1);
+			} else if (flinkConfiguration.containsKey(ConfigConstants.TASK_MANAGER_NUM_TASK_SLOTS)) {
+				slots = flinkConfiguration.getInteger(ConfigConstants.TASK_MANAGER_NUM_TASK_SLOTS, -1);
+			}
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: I think we should make the behaviour consistent with Mesos and standalone where we set `slots = flinkConfiguration.getInteger(ConfigConstants.TASK_MANAGER_NUM_TASK_SLOTS, 1)`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java [hunk 4]
```diff
@@ -335,8 +335,7 @@ protected void requestNewWorkers(int numWorkers) {
 			Priority priority = Priority.newInstance(0);
 
 			// Resource requirements for worker containers
-			int taskManagerSlots = taskManagerParameters.numSlots();
-			int vcores = config.getInteger(ConfigConstants.YARN_VCORES, Math.max(taskManagerSlots, 1));
+			int vcores = Math.max(taskManagerParameters.numSlots(), 1);
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java]: I think we should still keep the possibility open to configure the `vcores` explicitly.

**Comment by WangTaoTheTonic** [comment]
@tillrohrmann I still cannot get your point entirely. don't this three configs(`-s/-ys`, `yarn.containers.vcores` and `taskmanager.numberOfTaskSlots` mean same thing? Do they have difference in usage except priority?

**Comment by tillrohrmann** [comment]
If I'm not mistaken, then `yarn.containers.vcores` defines the number of vcores for a container. This, however, is not the same as the number of slots, since a slot can also have more than one vcore assigned.

**Comment by WangTaoTheTonic** [comment]
All right. That makes sense. Let me rephrase that and please check if we are in same channel:

1. slots of taskmanager is decided by `-s/-ys` and `taskmanager.numberOfTaskSlots`, the former has higher priority; and 
2. the vcores of yarn container is decided by `yarn.containers.vcores`, which will use values of `-s/-ys` or `taskmanager.numberOfTaskSlots` if user doesn't set `yarn.containers.vcores` explicitly.

Is that right?

**Comment by tillrohrmann** [comment]
Yes exactly @WangTaoTheTonic :-)

**Comment by WangTaoTheTonic** [comment]
@tillrohrmann 
After changing code the test results(both session and single mode) is like:

| Configurations                           | #vcores of container(TM) | #slots of TM |
| ---------------------------------------- | ------------------------ | ------------ |
| -s/-ys 5,  yarn.containers.vcores: 4,  taskmanager.numberOfTaskSlots: 3 | 4                        | 5            |
| yarn.containers.vcores: 4,  taskmanager.numberOfTaskSlots: 3 | 4                        | 3            |
| yarn.containers.vcores: 4                | 4                        | 1            |
| -s/-ys 5,  taskmanager.numberOfTaskSlots: 3 | 5                        | 5            |
| taskmanager.numberOfTaskSlots: 3         | 3                        | 3            |
| Nothing to specify, all use default      | 1                        | 1            |

Please check :)

**Comment by tillrohrmann** [comment]
Changes look good to me. Thanks for your contribution @WangTaoTheTonic. Merging this PR.
